toqito/state_props/is_separable.py

```diff
@@ -1,27 +1,71 @@
 
+EPS = np.finfo(float).eps
+
 
 def is_separable(state: np.ndarray, dim: None | int | list[int] = None, level: int = 2, tol: float = 1e-8) -> bool:
     r"""Determine if a given state (given as a density matrix) is a separable state :cite:`WikiSepSt`.
 
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :raises ValueError: If dimension is not specified correctly or input is invalid.
+    :param state: The density matrix to check.
+    :param dim: The dimension of the input state, e.g., [dim_A, dim_B].
+    :param level: The level for symmetric extensions.
+    :param tol: Numerical tolerance.
+    :return: True if separable, False if entangled or inconclusive by implemented checks.
+
+    Overview
+    ==========
+    :cite:`Horodecki2009_QuantumEntanglement_Review`
+    A quantum state :math:`\rho \in \text{D}(\mathcal{H}_A \otimes \mathcal{H}_B)` is called
+    separable if it can be written as a convex combination of product states.
+    If a state is not separable, it is called entangled.
+
+    This function implements several criteria:
+    1. Input validation (PSD, square, trace 1 normalization).
+    2. Trivial cases (e.g., 1D subsystem).
+    3. Pure states (Schmidt rank).
+    4. Gurvits-Barnum ball :cite:`Gurvits2002_LargestSeparableBalls`.
+    5. PPT criterion (Peres-Horodecki) :cite:`Peres1996_SeparabilityCriteria`,
+    :cite:`Horodecki1996_SeparabilityCriterion`.
+        - If PPT and :math:`MN \le 6`, it's separable :cite:`Horodecki1996_SeparabilityCriterion`.
+        - Operational criterion for low-rank PPT states (Horodecki et al. 2000
+        :cite:`Horodecki2000_OperationalCriterionLowRank`):
+            - If :math:`\text{rank}(\rho) \le \max(\text{dim}_A, \text{dim}_B)`.
+            - If :math:`\text{rank}(\rho) + \text{rank}(\rho^{T_A}) \le 2 MN - M - N + 2`.
+        - Range criterion variant for PPT states (Horodecki 1997 :cite:`Horodecki1997_RangeCriterion`).
+        - Rank-1 perturbation of identity for PPT states
+        (Vidal & Tarrach 1999 :cite:`Vidal1999_RobustnessOfEntanglement`).
+    6. 3x3 rank-4 PPT N&S check (e.g., Breuer 2005 :cite:`Breuer2005_VolumeOfTheSetOfSeparableStates`,
+      based on earlier work).
+    7. Reduction criterion (Horodecki 1999 :cite:`Horodecki1999_ReductionCriterion`).
+    8. Realignment/CCNR criteria (Chen & Wu 2003 :cite:`Chen2003_MatrixRealignment`,
+    Zhang et al. 2008 :cite:`Zhang2008_EntanglementDetectionBeyondRealignment`).
+    9. 2xN specific checks for PPT states (Johnston 2013 :cite:`Johnston2013_SeparabilityFromSpectrum`,
+    Hildebrand :cite:`Hildebrand2005_ConeOfPositiveMaps`).
+    10. Decomposable maps: Ha-Kye :cite:`Ha2011_PositiveMapsAndEntanglementQutrits`, Breuer-Hall
+    :cite:`Breuer2006_OptimalEntanglementCriterion`, :cite:`Hall2006_IndecomposablePositiveMaps`.
+    11. Symmetric extension hierarchy (DPS :cite:`Doherty2004_DistinguishingSeparableAndEntangledStates`).
```

Comment:
Check the formatting 

https://toqito--1236.org.readthedocs.build/en/1236/autoapi/state_props/is_separable/index.html